src/Lazy/Tree.elm

```diff
@@ -0,0 +1,618 @@
+    --> [ 2, 3 ]
+
+    singleton 1
+        |> insert (singleton 2)
+        |> item
+    --> 1
+
+-}
+insert : Tree a -> Tree a -> Tree a
+insert t (Tree treeItem c) =
+    Tree treeItem <| LL.append c <| LL.fromList [ t ]
+
+
+
+-- Forest
+
+
+{-| Construct `Forest` from a list.
+
+    import Lazy.LList as LL
+
+    [ { id = 1, parent = Nothing }
+    , { id = 2, parent = Nothing }
+    , { id = 3, parent = Just 1 }
+    ]
+        |> fromList (\p i -> Maybe.map .id p == i.parent)
+        |> LL.map (.id << item)
+        |> LL.toList
+    --> [ 1, 2 ]
+
+    [ { id = 1, parent = Nothing }
+    , { id = 2, parent = Nothing }
+    , { id = 3, parent = Just 1 }
+    , { id = 4, parent = Just 1 }
+    , { id = 5, parent = Just 2 }
+    ]
+        |> fromList (\p i -> Maybe.map .id p == i.parent)
+        |> LL.andThen descendants
+        |> LL.map (.id << item)
+        |> LL.toList
+    --> [ 3, 4, 5 ]
+
+-}
+fromList : (Maybe a -> a -> Bool) -> List a -> Forest a
+fromList isParent =
+    fromList_ Nothing isParent
+
+
+fromList_ : Maybe a -> (Maybe a -> a -> Bool) -> List a -> Forest a
+fromList_ parent isParent list =
+    LL.llist (List.map (constructTree isParent list) << List.filter (isParent parent)) list
+
+
+{-| Map function over `Forest`.
+
+    import Lazy.LList as LL
+
+    [ 1, 2, 3 ]
+        |> fromList (\m _ -> m == Nothing)
+        |> forestMap ((+) 1)
+        |> LL.map item
+        |> LL.toList
+    --> [ 2, 3, 4 ]
+
+-}
+forestMap : (a -> b) -> Forest a -> Forest b
+forestMap fc =
+    LL.map (map fc)
+
+
+{-| Map function over two `Forest`s.
+
+    import Lazy.LList as LL
+
+    [ 1, 2, 3 ]
+        |> fromList (\m _ -> m == Nothing)
+        |> forestMap2 (+) (fromList (\m _ -> m == Nothing) [1, 2])
+        |> LL.map item
+        |> LL.toList
+    --> [ 2, 4 ]
+
+-}
+forestMap2 : (a -> b -> c) -> Forest a -> Forest b -> Forest c
+forestMap2 fc =
+    LL.map2 (map2 fc)
+
+
+{-| Duplicates Tree (Comonad)
+-}
+duplicate : Tree a -> Tree (Tree a)
+duplicate ((Tree _ xs) as t) =
+    Tree t <| LL.map duplicate xs
+
+
+{-| Extend tree (Comonad)
+-}
+extend : (Tree a -> b) -> Tree a -> Tree b
+extend f =
+    map f << duplicate
+
+
+{-| Determine if any tree satisfy some test.
+
+This function is mutually recursing on the `isInForest` function.
+
+This function usually is the entry point of the mutual recursion,
+and specifies the behavior on the tree.
+
+    import Lazy.LList as LL
+
+    singleton 1
+        |> isInTree ((==) 1)
+    --> True
+
+    singleton 1
+        |> isInTree ((/=) 1)
+    --> False
+
+    Tree 1 (LL.fromList [ singleton 2, singleton 3, singleton 4 ])
+        |> isInTree ((==) 4)
+    --> True
+
+    Tree 1 (LL.fromList [ insert (singleton 5) <| singleton 2, insert (singleton 6) <| singleton 3, singleton 4 ])
+        |> isInTree ((==) 4)
+    --> True
+
+-}
+isInTree : (a -> Bool) -> Tree a -> Bool
+isInTree predicate (Tree value forest) =
+    predicate value || isInForest predicate forest
```

Comment:
Highlighting the additions to the vendored libraries 1/3

Reply:
Btw @annaghi do you have an idea when would this blow up the stack? Would we eg. have to have a tree `MAX_FN_CALLS` deep?

(I'm even fine with no answer. Just wondering about the potential limits of this recursion)